File: workout-zio/src/main/scala/workout/domain/WorkoutListResponse.scala
package workout.domain

import zio.json._

final case class WorkoutListResponse(
    success: Boolean,
    data: List[Workout],
    total: Int,
    message: Option[String]
)

object WorkoutListResponse {
  implicit val encoder: JsonEncoder[WorkoutListResponse] = DeriveJsonEncoder.gen[WorkoutListResponse]
  implicit val decoder: JsonDecoder[WorkoutListResponse] = DeriveJsonDecoder.gen[WorkoutListResponse]

  def success(workouts: List[Workout]): WorkoutListResponse =
    WorkoutListResponse(success = true, data = workouts, total = workouts.size, message = None)

  def error(msg: String): WorkoutListResponse =
    WorkoutListResponse(success = false, data = List.empty, total = 0, message = Some(msg))
}

File: workout-spring-webflux/src/main/java/com/workout/app/api/dto/WorkoutListResponse.java
package com.workout.app.api.dto;

import com.workout.app.domain.Workout;
import java.util.List;

public record WorkoutListResponse(boolean success, List<Workout> data, int total, String message) {
    public static WorkoutListResponse success(List<Workout> workouts) {
        return new WorkoutListResponse(true, workouts, workouts.size(), null);
    }
    public static WorkoutListResponse error(String msg) {
        return new WorkoutListResponse(false, List.of(), 0, msg);
    }
}

